File: commons/src/main/java/net/xpece/android/widget/XpNumberPicker.java
package net.xpece.android.widget;

import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.os.Build;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.drawable.DrawableCompat;
import android.widget.NumberPicker;

import java.lang.reflect.Field;

public final class XpNumberPicker {
    private static final Field FIELD_SELECTION_DIVIDER;

    static {
        Field f = null;
            try {
                f = NumberPicker.class.getDeclaredField("mSelectionDivider");
                f.setAccessible(true);
            } catch (NoSuchFieldException e) {
            }
        FIELD_SELECTION_DIVIDER = f;
    }

    private XpNumberPicker() {}

    public static void setSelectionDivider(@NonNull NumberPicker picker, @Nullable Drawable divider) {
        Drawable old = getSelectionDivider(picker);
        if (old != divider) {
            if (old != null) {
                old.setCallback(null);
            }
            try {
                FIELD_SELECTION_DIVIDER.set(picker, divider);
                if (divider != null) {
                    divider.setCallback(picker);
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        divider.setLayoutDirection(picker.getLayoutDirection());
                    }
                    if (divider.isStateful()) {
                        divider.setState(picker.getDrawableState());
                    }
                }
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
        }
    }

    @Nullable
    public static Drawable getSelectionDivider(@NonNull NumberPicker picker) {
        try {
            return (Drawable) FIELD_SELECTION_DIVIDER.get(picker);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static void setSelectionDividerTint(@NonNull NumberPicker picker, @Nullable ColorStateList color) {
        Drawable d = getSelectionDivider(picker);
        if (d != null) {
            d = DrawableCompat.wrap(d);
            DrawableCompat.setTintList(d, color);
            setSelectionDivider(picker, d);
        }
    }
}

File: commons/src/main/java/net/xpece/android/widget/XpNumberPicker.kt
@file:JvmName("XpNumberPicker")

package net.xpece.android.widget

import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.util.Log
import android.widget.NumberPicker
import androidx.core.graphics.drawable.DrawableCompat
import net.xpece.android.view.realLayoutDirection
import java.lang.reflect.Field

private object NumberPickerReflection {

    internal val selectionDividerField: Field? = try {
        NumberPicker::class.java
                .getDeclaredField("mSelectionDivider")
                .apply { isAccessible = true }
    } catch (ex: NoSuchFieldException) {
        Log.w("XpNumberPicker", "Couldn't fetch NumberPicker.mSelectionDivider field.", ex)
        null
    }

}

private var NumberPicker.selectionDivider: Drawable?
    get() {
        val f = NumberPickerReflection.selectionDividerField
        if (f == null) {
            Log.w("XpNumberPicker", "Couldn't get NumberPicker.mSelectionDivider field.")
            return null
        }
        return f.get(this) as Drawable?
    }
    set(value) {
        val old = selectionDivider
        if (old !== value) {
            val f = NumberPickerReflection.selectionDividerField
            if (f == null) {
                Log.w("XpNumberPicker", "Couldn't set NumberPicker.mSelectionDivider field.")
                return
            }
            if (old != null) {
                old.callback = null
            }
            f.set(this, value)
            if (value != null) {
                value.callback = this
                DrawableCompat.setLayoutDirection(value, realLayoutDirection)
                if (value.isStateful) {
                    value.state = drawableState
                }
            }
        }
    }

fun NumberPicker.setSelectionDividerTint(color: ColorStateList?) {
    var d = selectionDivider
    if (d != null) {
        d = DrawableCompat.wrap(d)
        DrawableCompat.setTintList(d, color)
        selectionDivider = d
    }
}
